### PR TITLE
Enhance fanout

### PR DESCRIPTION
**Pull request by iliacimpoes**

### CODE REVIEW

File: clientlibrary/partition/partition.go [hunk 1]
```diff
@@ -56,3 +56,15 @@ func (ss *ShardStatus) SetLeaseOwner(owner string) {
 	defer ss.Mux.Unlock()
 	ss.AssignedTo = owner
 }
+
+func (ss *ShardStatus) GetCheckpoint() string {
```
Comment (iliacimpoes) [inline clientlibrary/partition/partition.go:60]: Got race conditions
```
==================
WARNING: DATA RACE
Write at 0x00c0005589a0 by goroutine 112:
  github.com/vmware/vmware-go-kcl/clientlibrary/worker.(*RecordProcessorCheckpointer).Checkpoint()
      .../go/pkg/mod/github.com/vmware/vmware-go-kcl@v1.0.0/clientlibrary/worker/record-processor-checkpointer.go:71 +0x1fb
  main.(*fanoutWorker).ProcessRecords()
      .../fanout/poc.go:222 +0x39c
  github.com/vmware/vmware-go-kcl/clientlibrary/worker.(*ShardConsumer).getRecords()
      .../go/pkg/mod/github.com/vmware/vmware-go-kcl@v1.0.0/clientlibrary/worker/shard-consumer.go:232 +0x1382
  github.com/vmware/vmware-go-kcl/clientlibrary/worker.(*Worker).eventLoop.func1()
      .../go/pkg/mod/github.com/vmware/vmware-go-kcl@v1.0.0/clientlibrary/worker/worker.go:295 +0xcb
Previous read at 0x00c0005589a0 by goroutine 100:
  github.com/vmware/vmware-go-kcl/clientlibrary/worker.(*Worker).eventLoop()
      .../go/pkg/mod/github.com/vmware/vmware-go-kcl@v1.0.0/clientlibrary/worker/worker.go:250 +0x4b8
  github.com/vmware/vmware-go-kcl/clientlibrary/worker.(*Worker).Start.func1()
      .../go/pkg/mod/github.com/vmware/vmware-go-kcl@v1.0.0/clientlibrary/worker/worker.go:124 +0x86
Goroutine 112 (running) created at:
  github.com/vmware/vmware-go-kcl/clientlibrary/worker.(*Worker).eventLoop()
      .../go/pkg/mod/github.com/vmware/vmware-go-kcl@v1.0.0/clientlibrary/worker/worker.go:293 +0xea4
  github.com/vmware/vmware-go-kcl/clientlibrary/worker.(*Worker).Start.func1()
      .../go/pkg/mod/github.com/vmware/vmware-go-kcl@v1.0.0/clientlibrary/worker/worker.go:124 +0x86
Goroutine 100 (running) created at:
  github.com/vmware/vmware-go-kcl/clientlibrary/worker.(*Worker).Start()
      .../go/pkg/mod/github.com/vmware/vmware-go-kcl@v1.0.0/clientlibrary/worker/worker.go:121 +0x3e9
  main.(*app).start()
      .../fanout/poc.go:184 +0xfa4
  main.main.func2()
      .../fanout/poc.go:84 +0x6d
==================
```

File: clientlibrary/config/config.go [hunk 2]
```diff
@@ -271,18 +283,18 @@ func empty(s string) bool {
 	return len(strings.TrimSpace(s)) == 0
 }
 
-// checkIsValuePositive make sure the value is possitive.
+// checkIsValueNotEmpty make sure the value is not empty.
```
Comment (iaroslav-ciupin) [inline clientlibrary/config/config.go]:
```suggestion
// checkIsValueNotEmpty makes sure the value is not empty.
```

File: clientlibrary/config/config.go [hunk 3]
```diff
@@ -271,18 +283,18 @@ func empty(s string) bool {
...
 func checkIsValueNotEmpty(key string, value string) {
 	if empty(value) {
 		// There is no point to continue for incorrect configuration. Fail fast!
-		log.Panicf("Non-empty value exepected for %v, actual: %v", key, value)
+		log.Panicf("Non-empty value expected for %v, actual: %v", key, value)
 	}
 }
 
 // checkIsValuePositive make sure the value is possitive.
```
Comment (iaroslav-ciupin) [inline clientlibrary/config/config.go]:
```suggestion
// checkIsValuePositive makes sure the value is positive.
```

File: clientlibrary/worker/worker.go [hunk 4]
```diff
@@ -194,6 +199,32 @@ func (w *Worker) initialize() error {
...
+				time.Sleep(sleepDuration)
+				continue
+			}
+			if err != nil {
+				log.Errorf("Could not get consumer ARN: %v", err)
+				return err
+			}
+			w.consumerARN = consumerARN
+			break
+		}
```
Comment (iaroslav-ciupin) [inline clientlibrary/worker/worker.go]: can you extract this to some helper and refactor code to smth like
```go
if EnhancedFanOutEnabled {
  switch {
  case ConsumerARN != "":
  case ConsumerName != "":
    w.fetchConsumerARNWithRetry()
  default:  
    return errors.New("Missing enhanced fan-out consumer name or ARN")
  }
}
```

File: clientlibrary/worker/shard-consumer.go [hunk 5]
```diff
@@ -271,15 +183,73 @@ func (sc *ShardConsumer) getRecords(shard *par.ShardStatus) error {
 	}
 }
 
+// commonShardConsumer implements common functionality for regular and enhanced fan-out consumers
+type commonShardConsumer struct {
```
Comment (iaroslav-ciupin) [inline clientlibrary/worker/shard-consumer.go]: maybe it's worth moving it to a separate file?

File: test/worker_test.go [hunk 6]
```diff
@@ -171,6 +174,67 @@ func TestWorkerAssumeRole(t *testing.T) {
 	runTest(kclConfig, false, t)
 }
 
+func TestEnhancedFanOutConsumer(t *testing.T) {
+	// At miminal. use standard logrus logger
```
Comment (iaroslav-ciupin) [inline test/worker_test.go]:
```suggestion
	// At minimal, use standard logrus logger
```

**Review by iaroslav-ciupin — COMMENTED** [review]
overall LGTM, have you tested this with eec on proto0 ?

**Comment by iliacimpoes** [comment]
@iaroslav-ciupin, no. I wrote my own tests to make sure it work in different scenarios (resharding, service restart etc).
But that makes sense 👍 , I'll deploy and test it on proto.

**Comment by iliacimpoes** [comment]
tested on proto0, eec-admin with Enhanced Fan-Out  https://jenkins-build.ellationeng.cx-mgmt.com/job/eec-automated-test-proto0/147/
Look good.

File: clientlibrary/worker/worker.go [hunk 7]
```diff
@@ -194,6 +199,33 @@ func (w *Worker) initialize() error {
...
+	}
+
+	if w.kclConfig.EnableEnhancedFanOutConsumer && w.kclConfig.EnhancedFanOutConsumerARN == "" {
+		if w.kclConfig.EnhancedFanOutConsumerName == "" {
+			return errors.New("Missing enhanced fan-out consumer name")
+		}
+	}
+	if w.kclConfig.EnableEnhancedFanOutConsumer && w.kclConfig.EnhancedFanOutConsumerARN != "" {
+		w.consumerARN = w.kclConfig.EnhancedFanOutConsumerARN
+	}
```
Comment (iaroslav-ciupin) [inline clientlibrary/worker/worker.go]: remove?

**Pull request closed without merging**